test/unit-tests/Spectacle/Syntax/Logic.hs

```diff
@@ -0,0 +1,92 @@
+
+module Spectacle.Syntax.Logic
+  ( tests,
+  )
+where
+
+import Control.Monad
+import Data.Function ((&))
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Test.Tasty
+import Test.Tasty.Hedgehog
+
+import Language.Spectacle.Lang
+import Language.Spectacle.Syntax.Error
+import Language.Spectacle.Syntax.NonDet
+import Language.Spectacle.Syntax.Logic
+import Language.Spectacle.Exception.RuntimeException
+
+-- -------------------------------------------------------------------------------------------------
+
+prop_negationInvolute :: Property
+prop_negationInvolute = property do
+  let negated = runLogical (complement (complement (pure True)))
+  case negated of
+    Left exc -> do
+      footnoteShow exc
+      failure
+    Right x -> x === [True]
+
+evensForall ::
+  Members '[Logic, NonDet, Error RuntimeException] effs =>
+  [Int] ->
+  Lang ctx effs Int
+evensForall xs = forall xs (pure . even)
+
+oddsExists ::
+  Members '[Logic, NonDet, Error RuntimeException] effs =>
+  [Int] ->
+  Lang ctx effs Int
+oddsExists xs = exists xs (pure . odd)
+
+runLogical ::
+  Lang ctx '[Logic, NonDet, Error RuntimeException] a ->
+  Either RuntimeException [a]
+runLogical = runLang . runError . runNonDetA . runLogic
```

Comment:
Could you explain the ordering of these interpreters (or whatever you're calling them)? In particular, I'm curious about the order in which `NonDet` and `Error` effects are interpreted, since it sounds like they could interact?

Reply:
The ordering in which effects are interpreted is the same as typical monad transformers. Interpreting `Error` after `NonDet` will return `Left exc :: Either RuntimeError [a]` if any of the alternatives in `[]` raise an exception, which is the desired behavior.